### ISSUE TITLE
The reasoner should support loops

### ISSUE THREAD

**Issue by Sophietje** [TNO/knowledge-engine#382]
In GitLab by @barry.nouwt.tno.nl on May 12, 2022, 12:10

The current reasoner does not support loops. A loop is a (set of) rule(s) in which the consequent matches the antecedent. The unit test described here `eu.knowledge.engine.reasoner.TransitivityTest` shows an example of a single rule of which the consequent matches the antecedent. Such rules are quite important to express some of the Smart Homes & Energy domain knowledge. 

One example is the SAREF4Building's  [`hasSpace` property](https://saref.etsi.org/saref4bldg/v1.1.2/#s4bldg:BuildingSpace), which allows a `BuildingSpace` to contain `BuildingSpace`s itself which can contain more `BuildingSpace`s, etc. In such a scenario, whenever an inner `BuildingSpace` `contains` for example a particular `Device`, we would like this `Device` to also be `contain`ed by any outer `BuildingSpace`s. Loop support would allow the reasoner to infer this and thus recognize more interoperability for data exchange.

Currently, in scenario's with loops, the KE reasoner enters an infinite loop while planning the data exchange. During a physical session @wilco.wijbrandi.tno.nl, @han.kruiger.tno.nl and @barry.nouwt.tno.nl came up with a strategy/algorithm to let the reasoner deal with loops in a predictable manner. The figure below shows the algorithm:

![image](https://gitlab.inesctec.pt/interconnect/knowledge-engine/uploads/018f98448b768e4384c21d8b9882e3d4/image.png)

In short, when the backward chaining algorithm detects a loop it starts a forward chaining process (because backward chaining has difficulties in dealing with loops because you cannot know beforehand how many times you have to loop). This forward process will visit all the nodes within the loop precisely once and stores any binding sets on its way. After everyone has been visited, it determines whether another visit is necessary (based on if there is additional data inferred) and if so, it visits all the nodes within the loop again precisely once. This process repeats until no new data (i.e. bindings) are inferred and the process terminates. The resulting binding set of this forward chaining loop process is then used to finish the non-loop backward process.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 12, 2022, 12:34

created branch [`368-the-reasoner-should-support-loops`](/interconnect/knowledge-engine/-/compare/master...368-the-reasoner-should-support-loops) to address this issue

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Jun 7, 2022, 08:00

mentioned in commit 310e8fabb74a1e48c58a8bb04fea3931ff2f6255

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Jun 7, 2022, 08:03

marked this issue as related to #372
